tests/unit/test_charm.py

```diff
@@ -1,67 +1,395 @@
+#!/usr/bin/env python3
 # Copyright 2022 Canonical Ltd.
 # See LICENSE file for licensing details.
-#
-# Learn more about testing at: https://juju.is/docs/sdk/testing
 
+import json
 import unittest
-from unittest.mock import Mock
+from unittest.mock import Mock, PropertyMock, call, patch
 
-from ops.model import ActiveStatus
-from ops.testing import Harness
+import yaml
+from ops import testing
+from ops.model import ActiveStatus, BlockedStatus, WaitingStatus
 
-from charm import OperatorTemplateCharm
+from charm import AlertmanagerConfigurerOperatorCharm
 
+TEST_MULTITENANT_LABEL = "some_test_label"
+TEST_CONFIG = f"""options:
+  multitenant_label:
+    type: string
+    description: |
+      Alertmanager Configurer has been designed to support multiple tenants. In a multitenant
+      Alertmanager Configurer setup, each alert is first routed on the tenancy label, and then
+      the routing tree is distinct for each tenant.
+    default: {TEST_MULTITENANT_LABEL}
+  templates_file:
+    type: string
+    description: |
+      Alertmanager templates definition file content. All templates need to go into this single
+      config option. This config option will be ignored in case templates section is defined
+      in the Alertmanager config file (in which case, user is responsible for writing templates
+      files to the container).
+    default: ""
+"""
+TEST_ALERTMANAGER_CONFIG_FILE = "/test/rules/dir/config_file.yml"
+TEST_ALERTMANAGER_DEFAULT_CONFIG = """route:
+  receiver: test_receiver
+  group_by:
+  - alertname
+  group_wait: 1234s
+  group_interval: 4321s
+  repeat_interval: 1111h
+receivers:
+- name: test_receiver
+    """
 
-class TestCharm(unittest.TestCase):
+
+class TestAlertmanagerConfigurerOperatorCharm(unittest.TestCase):
+    @patch("charm.KubernetesServicePatch", lambda charm, ports: None)
     def setUp(self):
-        self.harness = Harness(OperatorTemplateCharm)
+        self.harness = testing.Harness(AlertmanagerConfigurerOperatorCharm, config=TEST_CONFIG)
         self.addCleanup(self.harness.cleanup)
+        testing.SIMULATE_CAN_CONNECT = True
+        self.harness.set_leader(True)
         self.harness.begin()
 
-    def test_config_changed(self):
-        self.assertEqual(list(self.harness.charm._stored.things), [])
-        self.harness.update_config({"thing": "foo"})
-        self.assertEqual(list(self.harness.charm._stored.things), ["foo"])
+    @patch("charm.AlertmanagerConfigDirWatcher")
+    @patch(
+        "charm.AlertmanagerConfigurerOperatorCharm.ALERTMANAGER_CONFIG_DIR",
+        new_callable=PropertyMock,
+    )
+    def test_given_alertmanager_config_directory_when_pebble_ready_then_watchdog_starts_watching_given_alertmanager_config_directory(  # noqa: E501
+        self, patched_config_dir, patched_alertmanager_config_dir_watcher
+    ):
+        test_config_dir = "/test/rules/dir"
+        patched_config_dir.return_value = test_config_dir
+        self.harness.charm.on.start.emit()
```

Comment:
Why are we using `self.harness.charm.on.start.emit()` instead of `self.harness.container_pebble_ready("alertmanager-configurer")`?

Reply:
Because watchdog starts on `start` event not on the `pebble-ready` event.